metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -1,264 +1,495 @@
 using mtg::common::logger;
 
 typedef AnnotatedDBG::node_index node_index;
 typedef AnnotatedDBG::row_index row_index;
 typedef AnnotatedDBG::Annotator::Label Label;
-typedef Alignment<DeBruijnGraph::node_index> DBGAlignment;
 
+typedef std::function<size_t()> LabelCountCallback;
 
-std::unique_ptr<bitmap_vector>
-mask_nodes_by_unitig(const DeBruijnGraph &graph,
-                     const KeepUnitigPath &keep_unitig,
-                     size_t num_threads) {
-    sdsl::bit_vector unitig_mask(graph.max_index() + 1, false);
+constexpr bool MAKE_BOSS = true;
 
-    const bool atomic = num_threads > 1;
 
-    graph.call_unitigs([&](const std::string &unitig, const auto &path) {
-        if (keep_unitig(unitig, path)) {
-            for (DeBruijnGraph::node_index node : path) {
-                set_bit(unitig_mask.data(), node, atomic);
-            }
-        }
-    }, num_threads);
+/**
+ * Return an int_vector<>, bitmap pair, each of length anno_graph.get_graph().get_base_graph().max_index().
+ * For an index i, the int_vector will contain a packed integer representing the
+ * number of labels in labels_in and labels_out which the k-mer of index i is
+ * annotated with. The least significant half of each integer represents the count
+ * from labels_in, while the most significant half represents the count from
+ * labels_out.
+ * The returned bitmap is a binarization of the int_vector
+ */
+std::pair<sdsl::int_vector<>, std::unique_ptr<bitmap>>
+construct_diff_label_count_vector(const AnnotatedDBG &anno_graph,
+                                  const std::vector<Label> &labels_in,
+                                  const std::vector<Label> &labels_out,
+                                  size_t num_threads);
+
+
+template <class GetKeptIntervals>
+void update_masked_graph_by_unitig(MaskedDeBruijnGraph &masked_graph,
+                                   const GetKeptIntervals &get_kept_intervals,
+                                   size_t num_threads);
+
+template <class KeepNode>
+void update_masked_graph_by_node(MaskedDeBruijnGraph &masked_graph,
+                                 const KeepNode &keep_node,
+                                 size_t num_threads);
+
+std::shared_ptr<MaskedDeBruijnGraph>
+make_initial_masked_graph(std::shared_ptr<const DeBruijnGraph> graph_ptr,
+                          sdsl::int_vector<> &counts,
+                          std::unique_ptr<bitmap>&& mask,
+                          bool add_complement,
+                          bool make_boss,
+                          size_t num_threads);
+
+
+MaskedDeBruijnGraph mask_nodes_by_label(const AnnotatedDBG &anno_graph,
+                                        const std::vector<Label> &labels_in,
+                                        const std::vector<Label> &labels_out,
+                                        const std::vector<Label> &labels_in_post,
+                                        const std::vector<Label> &labels_out_post,
```

Comment:
add a comment to explain what "_post" means

Reply:
I renamed `_post` to `_round2` and provided an explanation in the header for what this exactly means.